lib/Sema/SemaDecl.cpp

```diff
@@ -11213,28 +11213,107 @@ void Sema::ActOnFinishKNRParamDeclarations(Scope *S, Declarator &D,
   }
 }
 
-// ActOnBoundsDecl: handle a Checked C bounds declaration for a declarator.
+static void checkBoundsDeclWithTypeAnnotation(Sema &S, DeclaratorDecl *D,
+                                             InteropTypeBoundsAnnotation *Expr,
+                                             bool IsReturnBounds) {
+  assert(D != nullptr && Expr != nullptr && !Expr->isInvalid());
+
+  QualType DeclaredTy = D->getType();
+  QualType AnnotTy = Expr->getType();
+  QualType OriginalAnnotTy = AnnotTy;
+
+  if (DeclaredTy.isNull() || AnnotTy.isNull())
+    return;
+
+  if (IsReturnBounds) {
+    assert(DeclaredTy->isFunctionType());
+    if (const FunctionType *FuncTy = DeclaredTy->getAs<FunctionType>())
+      DeclaredTy = FuncTy->getReturnType();
+    else
+      return;
+  }
+
+  if (isa<ParmVarDecl>(D)) {
+    AnnotTy = S.Context.getAdjustedParameterType(AnnotTy);
+  }
+
+  int DiagId = 0;
+  // Check that declared type is allowed to have a bounds-safe
+  // interface type declaration.
+  if (!DeclaredTy->isPointerType() && !DeclaredTy->isArrayType()) {
+    if (IsReturnBounds)
+      DiagId =
+        diag::err_typecheck_return_bounds_type_annotation_for_illegal_type;
+    else
+      DiagId = diag::err_typecheck_bounds_type_annotation_for_illegal_type;
+    S.Diag(Expr->getStartLoc(), DiagId);
+    return;
+  }
+
+   // Check that the annotation type is allowed to appear in a bound-safe
+  // interface type declaration.
+  SourceLocation AnnotTyLoc =
+    Expr->getTypeInfoAsWritten()->getTypeLoc().getBeginLoc();
+
+  if (IsReturnBounds) {
+    if (AnnotTy->isArrayType())
+      DiagId = diag::err_typecheck_return_bounds_type_annotation_is_array;
+    else if (!AnnotTy->isPointerType())
+      DiagId =
+        diag::err_typecheck_return_bounds_type_annotation_must_be_pointer;
+  } else {
+    if (!AnnotTy->isPointerType() && !AnnotTy->isArrayType())
+      DiagId = diag::err_typecheck_bounds_type_annotation_must_be_pointer_or_array;
+  }
+
+  if (DiagId) {
+    S.Diag(AnnotTyLoc, DiagId);
+    return;
+  }
+
+  // Check that the types are identical if checking is ignored.
+  if (!S.Context.isEqualIgnoringChecked(AnnotTy, DeclaredTy)) {
+    DiagId = diag::err_typecheck_bounds_type_annotation_incompatible;
+    S.Diag(AnnotTyLoc, DiagId) << AnnotTy << DeclaredTy;
+    return;
```

Comment:
Yes, it makes sense to make a "best effort" to generate multiple distinct errors at once.    This could be a little tircky beause we need to void generating errors that are implied by other errors, though.  For example, if a bounds-afe interface type isn't an array or pointer type, there's no point in complaining about it being unchecked.